app/Libraries/ModdingHistoryEventsBundle.php

```diff
@@ -0,0 +1,274 @@
+                    'statistics.rank',
+                    'support_level',
+                    'unranked_beatmapset_count',
+                ]
+            );
+        }
+
+        return $array;
+    }
+
+    private function getDiscussions()
+    {
+        $includes = [
+            'beatmap',
+            'beatmapDiscussionVotes',
+            'beatmapset',
+            'startingPost',
+        ];
+
+        $parents = BeatmapDiscussion::search($this->searchParams);
+        $parents['query']->with($includes);
+
+        if ($this->isModerator) {
+            $parents['query']->visibleWithTrashed();
+        } else {
+            $parents['query']->visible();
+        }
+
+        $discussions = $parents['query']->get();
+
+        // TODO: remove this when reviews are released
+        if (!config('osu.beatmapset.discussion_review_enabled')) {
+            return $discussions;
+        }
+
+        $children = BeatmapDiscussion::whereIn('parent_id', $discussions->pluck('id'))->with($includes);
+
+        if ($this->isModerator) {
+            $children->visibleWithTrashed();
+        } else {
+            $children->visible();
+        }
+
+        return $discussions->merge($children->get());
+    }
+
+    private function getEvents()
+    {
+        $events = BeatmapsetEvent::search($this->searchParams);
+        $events['query'] = $events['query']->with([
+            'beatmapset.user',
+            'beatmapDiscussion.beatmapset',
+            'beatmapDiscussion.startingPost',
+        ])->whereHas('beatmapset');
+
+        if ($this->isModerator) {
+            $events['query']->with(['beatmapset' => function ($query) {
+                $query->withTrashed();
+            }]);
+        }
+
+        // just for the paginator
+        $this->total = $events['query']->realCount();
+        $this->params = $events['params'];
+
+        return $events['query']->get();
+    }
+
+    private function getPosts()
+    {
+        $posts = BeatmapDiscussionPost::search($this->searchParams);
+        $posts['query']->with([
+            'beatmapDiscussion.beatmap',
+            'beatmapDiscussion.beatmapset',
+        ]);
+
+        if (!$this->isModerator) {
+            $posts['query']->visible();
+        }
+
+        return $posts['query']->get();
+    }
+
+    private function getUsers()
+    {
+        $userIds = [];
+        foreach ($this->discussions as $discussion) {
+            $userIds[] = $discussion->user_id;
+            $userIds[] = $discussion->startingPost->last_editor_id;
+        }
+
+        $votesGiven = ($this->votes['given'] ?? collect())->pluck('user_id')->toArray();
+        $votesReceived = ($this->votes['received'] ?? collect())->pluck('user_id')->toArray();
+
+        $userIds = array_merge(
+            $userIds,
+            $this->posts->pluck('user_id')->toArray(),
+            $this->posts->pluck('last_editor_id')->toArray(),
+            $this->events->pluck('user_id')->toArray(),
```

Comment:
looks like bug waiting to happen when someone moves `getUsers` call to before those are assigned (also `$this->votes` and `$this->discussions` above).